Updated handling of ICollectionElementInitializerOperation arguments

333fred opened the pull request:
We now generate IArgumentOperations for the arguments to collection initializers, including the arguments that were previously skipped over such as default arguments. Additionally, we break out dynamic collection element initializers into their own IOperation, as we cannot determine at compile-time what add method is being called.

Tagging @dotnet/roslyn-compiler @AlekseyTs @dotnet/analyzer-ioperation for review.

Edit:
Fixes https://github.com/dotnet/roslyn/issues/18550.

jaredpar commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -1,3 +1,4 @@
+*REMOVED*Microsoft.CodeAnalysis.Operations.ICollectionElementInitializerOperation.Arguments.get -> System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.IOperation>
```
>OVED [](start = 4, length = 4)

Removal of a shipped API generally requires compat council approval. Is there a different bar here?

333fred replied:
API is not getting removed, no longer relevant.

---
In reply to: [187511037](https://github.com/dotnet/roslyn/pull/26781#discussion_r187511037) [](ancestors = 187511037)

jaredpar commented on src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:
```diff
@@ -28,11 +27,16 @@ public interface ICollectionElementInitializerOperation : IOperation
+        /// <see cref="IDynamicCollectionElementInitializerOperation"/>.
+        /// </remarks>
+        [Obsolete]
         bool IsDynamic { get; }
```
This PR appears to remove one API and Obsolete another. Can you explain why?

333fred replied:
As mentioned in the email thread, I'm currently following the design that seemed to have the most support. This is still subject to design team approval and compat council approval.

jaredpar commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -4640,7 +4640,12 @@ public override IOperation VisitAnonymousObjectCreation(IAnonymousObjectCreation
+
+        public override IOperation VisitDynamicCollectionElementInitializer(IDynamicCollectionElementInitializerOperation operation, int? argument)
+        {
+            return new DynamicCollectionElementInitializerOperation(VisitArray(operation.Arguments), ((Operation)operation).SemanticModel, operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation));
```
Why does this get a semantic model bot non-dynamic collection initializer does not?

333fred replied:
No longer relevant, but was a bug.

---
In reply to: [187511369](https://github.com/dotnet/roslyn/pull/26781#discussion_r187511369) [](ancestors = 187511369)

333fred commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ObjectOrCollectionInitializerExpression.cs:
```diff
@@ -171,10 +171,10 @@ private BoundExpression MakeCollectionInitializer(BoundExpression rewrittenRecei
             }
 
-            return MakeCall(null, syntax, rewrittenReceiver, addMethod, rewrittenArguments, default(ImmutableArray<RefKind>), initializer.InvokedAsExtensionMethod, initializer.ResultKind, addMethod.ReturnType, temps);
+            return MakeCall(null, syntax, rewrittenReceiver, addMethod, rewrittenArguments, argumentRefKindsOpt, initializer.InvokedAsExtensionMethod, initializer.ResultKind, addMethod.ReturnType, temps);
```
Will revert, didn't intend to commit this change.

jaredpar commented on src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs:
```diff
@@ -878,11 +878,20 @@ public override void VisitCollectionElementInitializer(ICollectionElementInitial
             var addMethod = operation.AddMethod;
+#pragma warning disable CS0612 // Type or member is obsolete
+            // Still making sure it doesn't crash
             var isDynamic = operation.IsDynamic;
```
>  var isDynamic = operation.IsDynamic; [](start = 10, length = 38)

I continue to find this pattern confusing. If we are calling code to assert that it does not throw that is reasonable but I think we need to make it more first class. As is currently being written if I came across the code in a few years I may delete it because it looks unnecessary. Consider adding an assert method of the form

```
void AssertNoThrow(Action action);
```

I understand xunit removed this because of their principaled stance on the matter. At the same time though it makes the code far more readable and their stance simply doesn't account for cases like this.

AlekseyTs replied:
> Consider adding an assert method of the form ...

It feels like allocating a delegate just for this purpose is an overkill. Perhaps just a generic method that takes ```T```.

```
void AssertNoThrow<T>(T val) {}
```

---
In reply to: [187724762](https://github.com/dotnet/roslyn/pull/26781#discussion_r187724762) [](ancestors = 187724762)

jaredpar replied:
That works for me as well.

---
In reply to: [187731690](https://github.com/dotnet/roslyn/pull/26781#discussion_r187731690) [](ancestors = 187731690,187724762)

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/26848](https://github.com/dotnet/roslyn/issues/26848) to follow up on this, as it's well beyond the scope of this PR.

---
In reply to: [187733960](https://github.com/dotnet/roslyn/pull/26781#discussion_r187733960) [](ancestors = 187733960,187731690,187724762)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:
```diff
@@ -28,11 +27,16 @@ public interface ICollectionElementInitializerOperation : IOperation
+        /// This flag is obsolete, and dynamic collection initializers have been broken out into a separate interface,
+        /// <see cref="IDynamicCollectionElementInitializerOperation"/>.
+        /// </remarks>
+        [Obsolete]
```
>[Obsolete] [](start = 8, length = 10)

Alternatively we could keep this node as is. It feels like splitting it is a major API change. We we unable to represent some dynamic cases before?

333fred replied:
No, we were not missing any dynamic cases before that we're getting with the new API

---
In reply to: [187732590](https://github.com/dotnet/roslyn/pull/26781#discussion_r187732590) [](ancestors = 187732590)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:
```diff
@@ -8,10 +9,8 @@ namespace Microsoft.CodeAnalysis.Operations
     /// Current usage:
-    ///  (1) C# nested collection element initializer expression within a collection initializer.
-    ///      For example, given a collection initializer "new Class() { Y = { { x, y, 3 } } }",
-    ///      nested collection element initializer for Y, i.e. "{ { x, y, 3 } }" is represented by this operation.
```
>nested collection element initializer for Y, i.e. "{ { x, y, 3 } }" is represented by this operation. [](start = 13, length = 101)

I think this example was valuable. Consider preserving it

AlekseyTs commented:
>         /// Add method invoked on collection. Null for dynamic invocation and error cases.

I assume with the current approach (splitting the node), this portion of the comment is obsolete and  ```AddMethod``` is never null.

---
Refers to: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:23 in 38c141b. [](commit_id = 38c141b83940a47fe09963b70c84988b0afc1fec, deletion_comment = False)

AlekseyTs commented:
>     public interface ICollectionElementInitializerOperation : IOperation

With the current approach (splitting the node), it might be appropriate for this node to implement IMemberReferenceOperation

---
Refers to: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:20 in 38c141b. [](commit_id = 38c141b83940a47fe09963b70c84988b0afc1fec, deletion_comment = False)

AlekseyTs commented:
>     /// Represents a dynamically bound new/New expression.

It looks like this portion of the comment is no longer accurate

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:5166 in 38c141b. [](commit_id = 38c141b83940a47fe09963b70c84988b0afc1fec, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -5260,6 +5260,55 @@ internal sealed partial class LazyDynamicObjectCreationExpression : BaseDynamicO
         public override IObjectOrCollectionInitializerOperation Initializer => SetParentOperation(_lazyInitializer.Value, this);
     }
 
+    internal abstract partial class BaseDynamicCollectionElementInitializerOperation : HasDynamicArgumentsExpression, IDynamicCollectionElementInitializerOperation
```
>HasDynamicArgumentsExpression, IDynamicCollectionElementInitializerOperation [](start = 87, length = 76)

It looks like for other interfaces that are implemented by classes inheriting from HasDynamicArgumentsExpression, we have extension methods that provide access to the the things stored in the base. 
For example:
```
        public static RefKind? GetArgumentRefKind(this IDynamicInvocationOperation dynamicOperation, int index)
        {
            if (dynamicOperation == null)
            {
                throw new ArgumentNullException(nameof(dynamicOperation));
            }

            return GetArgumentRefKind((HasDynamicArgumentsExpression)dynamicOperation, index);
        }
```


Are we going to provide similar extension methods for IDynamicCollectionElementInitializerOperation? Is there any useful information to expose? If not, then consider saving some memory, inheriting from Operation directly and defining ```public abstract ImmutableArray<IOperation> Arguments { get; }```` in this class directly.

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -5260,6 +5260,55 @@ internal sealed partial class LazyDynamicObjectCreationExpression : BaseDynamicO
+        public override ImmutableArray<IOperation> Arguments { get; }
+    }
+
+    internal sealed partial class LazyDynamicCollectionElemementInitializerOperation : BaseDynamicCollectionElementInitializerOperation
```
>Elemement [](start = 55, length = 9)

Typo "Elemement"?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -699,13 +699,25 @@ private IOperation CreateBoundObjectInitializerMemberOperation(BoundObjectInitia
-            Lazy<ImmutableArray<IOperation>> arguments = new Lazy<ImmutableArray<IOperation>>(() => boundCollectionElementInitializer.Arguments.SelectAsArray(n => Create(n)));
-            bool isDynamic = false;
+            Lazy<ImmutableArray<IArgumentOperation>> arguments = new Lazy<ImmutableArray<IArgumentOperation>>(() =>
+                DeriveArguments(
```
>DeriveArguments [](start = 16, length = 15)

It looks like the information in BoundCollectionElementInitializer is actually getting pulled from BoundCall node and in some cases compiler can create BoundCall node even for erroneous cases. See how CreateBoundCallOperation helper is trying to intercept those, failure to do so is likely to cause a crash in DeriveArguments. I think we should be conservative and add similar guard here, come up with a test scenario would be good too.

333fred replied:
CreateBoundCall is trying to compensate for overload resolution failure, which is not a case we'll ever see in this method. See BindCollectionInitializerElementAddMethod in Binder_Expressions.cs. If overload resolution failed, we'd end up with a BoundCall, which will be properly handled by that create method.

---
In reply to: [187750757](https://github.com/dotnet/roslyn/pull/26781#discussion_r187750757) [](ancestors = 187750757)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:
```diff
@@ -28,11 +27,16 @@ public interface ICollectionElementInitializerOperation : IOperation
         /// Arguments passed to add method invocation.
         /// </summary>
-        ImmutableArray<IOperation> Arguments { get; }
+        ImmutableArray<IArgumentOperation> Arguments { get; }
```
>ImmutableArray<IArgumentOperation> Arguments { get; } [](start = 8, length = 53)

Should we also add ```bool IsVirtual { get; }``` property?

jaredpar replied:
What would be the value in the IsVirtual property? 

---
In reply to: [187751117](https://github.com/dotnet/roslyn/pull/26781#discussion_r187751117) [](ancestors = 187751117)

AlekseyTs replied:
> What would be the value in the IsVirtual property? 

The same value as in IInvocationOperation.IsVirtual property - makes it clear whether the call is virtual or not. 

---
In reply to: [188026663](https://github.com/dotnet/roslyn/pull/26781#discussion_r188026663) [](ancestors = 188026663,187751117)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -783,6 +819,129 @@ static void Main()
+            ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 3, IsInvalid) (Syntax: '3')
+";
+            var expectedDiagnostics = new DiagnosticDescription[] {
+                Diagnostic(ErrorCode.ERR_InitializerAddHasParamModifiers, "1").WithArguments("C.Add(ref int)").WithLocation(10, 35),
```
>Diagnostic(ErrorCode.ERR_InitializerAddHasParamModifiers, "1").WithArguments("C.Add(ref int)").WithLocation(10, 3 [](start = 16, length = 113)

Please include the actual message and the actual source code for the line as we usually do when we assert diagnostics

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -783,6 +819,129 @@ static void Main()
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void ObjectCreationWithDefaultParameterAddMethod()
```
>ObjectCreationWithDefaultParameterAddMethod [](start = 20, length = 43)

Is ```params``` case interesting too?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -4453,6 +4453,7 @@ private BoundExpression BindUnexpectedComplexElementInitializer(InitializerExpre
                     boundCall.ArgsToParamsOpt,
                     boundCall.InvokedAsExtensionMethod,
                     boundCall.ResultKind,
+                    binderOpt: this,
```
>this [](start = 31, length = 4)

```boundCall.BinderOpt```?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.vb:
```diff
@@ -540,6 +564,137 @@ BC30456: 'MissingField' is not a member of 'C'.
+
+        <CompilerTrait(CompilerFeature.IOperation)>
+        <Fact, WorkItem(22967, "https://github.com/dotnet/roslyn/issues/22967")>
+        Public Sub ObjectCreationWithOptionalParameterAddMethod()
```
>ObjectCreationWithOptionalParameterAddMethod [](start = 19, length = 44)

ParamArray?

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs:
```diff
@@ -878,11 +878,20 @@ public override void VisitCollectionElementInitializer(ICollectionElementInitial
         {
             Assert.Equal(OperationKind.CollectionElementInitializer, operation.Kind);
             var addMethod = operation.AddMethod;
```
>operation.AddMethod [](start = 28, length = 19)

Assert.NotNull?

AlekseyTs commented:
Done with review pass (iteration 2)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs:
```diff
@@ -3294,15 +3348,24 @@ static int Main()
               IObjectOrCollectionInitializerOperation (OperationKind.ObjectOrCollectionInitializer, Type: A) (Syntax: '{ 4, 5, 6 }')
                 Initializers(3):
-                    ICollectionElementInitializerOperation (AddMethod: void A.Add(System.Int32 i)) (IsDynamic: False) (OperationKind.CollectionElementInitializer, Type: System.Void, IsImplicit) (Syntax: '4')
+                    ICollectionElementInitializerOperation (AddMethod: void A.Add(System.Int32 i)) (OperationKind.CollectionElementInitializer, Type: System.Void, IsImplicit) (Syntax: '4')
```
>void A.Add(System.Int32 i) [](start = 71, length = 26)

Do we have any tests covering "Add" extension methods. It would be good to have some. Especially for C#, it is interesting to see what happens with the implicit receiver.

AlekseyTs commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -5260,6 +5260,55 @@ internal sealed partial class LazyDynamicObjectCreationExpression : BaseDynamicO
+    internal abstract partial class BaseDynamicCollectionElementInitializerOperation : HasDynamicArgumentsExpression, IDynamicCollectionElementInitializerOperation
+    {
+        protected BaseDynamicCollectionElementInitializerOperation(SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) :
+            base(OperationKind.DynamicCollectionElementInitializer, argumentNames: default, argumentRefKinds: default, semanticModel, syntax, type, constantValue, isImplicit)
```
>argumentRefKinds: default [](start = 92, length = 25)

I am not sure if default accurately reflects semantics for C#. Should the invocation force non-ref nature of the parameters?

333fred commented:
>         /// Add method invoked on collection. Null for dynamic invocation and error cases.

Correct

---
In reply to: [388485797](https://github.com/dotnet/roslyn/pull/26781#issuecomment-388485797) [](ancestors = 388485797)

---
Refers to: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:23 in 38c141b. [](commit_id = 38c141b83940a47fe09963b70c84988b0afc1fec, deletion_comment = False)

333fred commented:
Another suggestion on potential API shape by @AlekseyTs offline:

> We should also consider deprecating the ICollectionElementInitializerOperation node altogether and instead use existing IInvocationOperation and IDynamicInvocationOperation nodes to represent the initializers. In similar cases for member initializers we don’t invent new nodes, just use ISimpleAssignmentOperation.

jaredpar commented:
@333fred is that proposal your current POR? Wanting to know if I should hold off on reviewing until that is submitted.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:
```diff
@@ -1,23 +1,23 @@
     /// This interface is reserved for implementation by its associated APIs. We reserve the right to
     /// change it in the future.
     /// </remarks>
+    [Obsolete]
```
>[Obsolete] [](start = 4, length = 10)

I am not sure what is our policy about marking existing APIs as obsolete. Might want to check with Compatibility Council

333fred replied:
Still figuring this out, has not yet been addressed.

---
In reply to: [189329437](https://github.com/dotnet/roslyn/pull/26781#discussion_r189329437) [](ancestors = 189329437)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationKind.cs:
```diff
@@ -112,7 +113,7 @@ public enum OperationKind
         MemberInitializer = 0x33,
         /// <summary>Indicates an <see cref="ICollectionElementInitializerOperation"/>.</summary>
-        CollectionElementInitializer = 0x34,
+        [Obsolete] CollectionElementInitializer = 0x34,
```
>[Obsolete] [](start = 8, length = 10)

Same concern about the attribute

AlekseyTs replied:
>[Obsolete]  [](start = 8, length = 11)

What about this attribute? Should it be an error?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationVisitor.cs:
```diff
@@ -362,6 +363,7 @@ public virtual void VisitMemberInitializer(IMemberInitializerOperation operation
             DefaultVisit(operation);
         }
 
+        [Obsolete]
```
>[Obsolete] [](start = 8, length = 10)

And here

AlekseyTs replied:
>[Obsolete] [](start = 8, length = 10)

What about this attribute? Should it be an error?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/OperationVisitor.cs:
```diff
@@ -909,6 +911,7 @@ public virtual TResult VisitMemberInitializer(IMemberInitializerOperation operat
             return DefaultVisit(operation, argument);
         }
 
+        [Obsolete]
```
>[Obsolete] [](start = 8, length = 10)

And here

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -4445,14 +4447,17 @@ private BoundExpression BindUnexpectedComplexElementInitializer(InitializerExpre
                     elementInitializer,
                     boundCall.Method,
                     boundCall.Arguments,
+                    invokedAsExtensionMethod ? null : implicitReceiver,
```
>invokedAsExtensionMethod ? null : implicitReceiver [](start = 20, length = 50)

It feels like rather than having this logic we should simply pull receiver from the boundCall, it should already be properly adjusted

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -696,16 +693,48 @@ private IOperation CreateBoundObjectInitializerMemberOperation(BoundObjectInitia
+            }
+            else
+            {
+                receiver = CreateReceiverOperation(boundCollectionElementInitializer.ImplicitReceiverOpt, addMethod);
```
>CreateReceiverOperation [](start = 27, length = 23)

Do we have a test covering static Add method?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -696,16 +693,48 @@ private IOperation CreateBoundObjectInitializerMemberOperation(BoundObjectInitia
+                Debug.Assert(boundCollectionElementInitializer.Arguments.Length >= 1);
+                Debug.Assert(boundCollectionElementInitializer.Arguments[0].Kind == BoundKind.ImplicitReceiver);
+                Debug.Assert(boundCollectionElementInitializer.ImplicitReceiverOpt == null);
+                receiver = OperationFactory.NullOperation;
```
>receiver = OperationFactory.NullOperation; [](start = 16, length = 42)

I am not sure why do we have this special case here, it looks like CreateReceiverOperation is able to handle this case just fine for the BoundCall node.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -737,16 +766,15 @@ private IDynamicMemberReferenceOperation CreateBoundDynamicMemberAccessOperation
             Optional<object> constantValue = ConvertToOptional(boundCollectionElementInitializer.ConstantValue);
             bool isImplicit = boundCollectionElementInitializer.WasCompilerGenerated;
-            return new LazyCollectionElementInitializerExpression(addMethod, isDynamic, arguments, _semanticModel, syntax, type, constantValue, isImplicit);
+            return new LazyDynamicInvocationExpression(operation, arguments, argumentNames: ImmutableArray<string>.Empty, argumentRefKinds: ImmutableArray<RefKind>.Empty, _semanticModel, syntax, type, constantValue, isImplicit);
```
>ImmutableArray<RefKind>.Empty [](start = 140, length = 29)

Is this going to accurately reflect that none of the arguments have ref/out/in modifiers?

333fred replied:
That's correct. All our current IDynamicInvocationOperation tests have an empty refkind/name array when there are none.

---
In reply to: [189336842](https://github.com/dotnet/roslyn/pull/26781#discussion_r189336842) [](ancestors = 189336842)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.vb:
```diff
@@ -540,6 +578,297 @@ BC30456: 'MissingField' is not a member of 'C'.
+              ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
+              InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+              OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+            IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: o) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: '1')
```
>ArgumentKind.Explicit [](start = 32, length = 21)

Should this be DefaultValue?

333fred replied:
This and below required plumbing explicit knowledge of default parameters through the binder. Change will be pushed later today.

---
In reply to: [189342049](https://github.com/dotnet/roslyn/pull/26781#discussion_r189342049) [](ancestors = 189342049)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.vb:
```diff
@@ -540,6 +578,297 @@ BC30456: 'MissingField' is not a member of 'C'.
+        Instance Receiver: 
+          IInstanceReferenceOperation (ReferenceKind: ImplicitReceiver) (OperationKind.InstanceReference, Type: Mod1.C, IsImplicit) (Syntax: 'New C From {1, 2, 3}')
+        Arguments(1):
+            IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: ints) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: '1')
```
>ArgumentKind.Explicit [](start = 32, length = 21)

Should this be ParamArray?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.vb:
```diff
@@ -540,6 +578,297 @@ BC30456: 'MissingField' is not a member of 'C'.
+]]>.Value
+
+            VerifyOperationTreeAndDiagnosticsForTest(Of ObjectCollectionInitializerSyntax)(source, expectedOperationTree, expectedDiagnostics)
+        End Sub
```
>End Sub [](start = 8, length = 7)

Do we have any tests for initializers for interface types? It would be good to confirm types of implicit receivers for cases like that.

AlekseyTs replied:
ObjectCreationCollectionInitializerExtensionAddMethodOnInterface does not test Add method on an interface, we would have to new an interface for that scenario. Should have similar test for C#

---
In reply to: [189343625](https://github.com/dotnet/roslyn/pull/26781#discussion_r189343625) [](ancestors = 189343625)

333fred replied:
Added ObjectCreationCollectionInitializerAddMethodOnInterface for both C# and VB. COM never ceases to amaze me.

---
In reply to: [190036238](https://github.com/dotnet/roslyn/pull/26781#discussion_r190036238) [](ancestors = 190036238,189343625)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -727,30 +728,14 @@ Namespace Microsoft.CodeAnalysis.Operations
             Return New LazyObjectOrCollectionInitializerExpression(initializers, _semanticModel, syntax, type, constantValue, isImplicit)
         End Function
 
-        Private Function CreateBoundCollectionElementInitializerOperation(boundExpression As BoundExpression) As IOperation
```
>CreateBoundCollectionElementInitializerOperation [](start = 25, length = 48)

Do we have any tests covering late bound invocation of the Add method? Didn't notice any among changed tests, but I could miss something.

333fred replied:
Adding one, but it is blocked by [https://github.com/dotnet/roslyn/issues/27034](https://github.com/dotnet/roslyn/issues/27034). Will fix in a separate PR.

---
In reply to: [189345284](https://github.com/dotnet/roslyn/pull/26781#discussion_r189345284) [](ancestors = 189345284)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -783,6 +841,293 @@ static void Main()
+              InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+              OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+            IArgumentOperation (ArgumentKind.DefaultValue, Matching Parameter: o) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: '1')
+              ILiteralOperation (OperationKind.Literal, Type: System.Object, Constant: null, IsImplicit) (Syntax: '1')
```
>ILiteralOperation (OperationKind.Literal, Type: System.Object, Constant: null, IsImplicit) (Syntax: '1') [](start = 14, length = 104)

This doesn't look right.

AlekseyTs commented:
Done with review pass (iteration 3)

333fred commented:
@AlekseyTs addressed feedback. Still figuring out the compat council for the Obsolete attribute.

---
In reply to: [390304176](https://github.com/dotnet/roslyn/pull/26781#issuecomment-390304176) [](ancestors = 390304176)

AlekseyTs commented:
@333fred It looks like CI builds are broken due to the obsolete attribute that you added:

error CS0618: 'ICollectionElementInitializerOperation' is obsolete: 'ICollectionElementInitializerOperation has been replaced with IInvocationOperation and IDynamicInvocationOperation' [D:\j\workspace\windows_relea---19dd66ba\src\Compilers\Core\Portable\CodeAnalysis.csproj]

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -642,6 +642,8 @@
         <Field Name="ReceiverOpt" Type="BoundExpression" Null="allow"/>
         <Field Name="Arguments" Type="ImmutableArray(Of BoundExpression)"/>
+        <!-- Which arguments are default. Used by IOperation for determining ArgumentKind. -->
+        <Field Name="DefaultArguments" Type="BitVector" Null="NotApplicable" />
```
>BitVector [](start = 45, length = 9)

We list all the value types used at the beginning of the root node. I do not remember exact details, but that probably affects how the classes are generated

333fred replied:
Ah, I missed that. When I was looking at how ValueTypes were generated I modified the generator, I'll take this approach instead.

---
In reply to: [190011872](https://github.com/dotnet/roslyn/pull/26781#discussion_r190011872) [](ancestors = 190011872)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundCall.vb:
```diff
@@ -7,6 +7,26 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            methodGroupOpt As BoundMethodGroup,
+            receiverOpt As BoundExpression,
+            arguments As ImmutableArray(Of BoundExpression),
+            defaultArguments As BitVector,
```
>defaultArguments As BitVector, [](start = 12, length = 30)

Why not simply add it as optional to the constructor below?

AlekseyTs replied:
Adding as non optional is even better, this way we will make sure that all call sites properly preserve information

---
In reply to: [190017640](https://github.com/dotnet/roslyn/pull/26781#discussion_r190017640) [](ancestors = 190017640)

333fred replied:
Adding as optional to the constructor below, as we have 69 callsites that don't pass DefaultArguments currently.

---
In reply to: [190018044](https://github.com/dotnet/roslyn/pull/26781#discussion_r190018044) [](ancestors = 190018044,190017640)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundCall.vb:
```diff
@@ -19,13 +39,31 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            Me.New(syntax, method, methodGroupOpt, receiverOpt, arguments, defaultArguments:=BitVector.Null, constantValueOpt, isLValue, suppressObjectClone, type, hasErrors)
+        End Sub
+
+        Public Function Update(method As MethodSymbol, methodGroupOpt As BoundMethodGroup, receiverOpt As BoundExpression, arguments As ImmutableArray(Of BoundExpression), constantValueOpt As ConstantValue, isLValue As Boolean, suppressObjectClone As Boolean, type As TypeSymbol) As BoundCall
```
>Public Function Update(meth [](start = 8, length = 27)

Please break this long line

333fred replied:
No longer relevant.

---
In reply to: [190018757](https://github.com/dotnet/roslyn/pull/26781#discussion_r190018757) [](ancestors = 190018757)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundObjectCreationExpression.vb:
```diff
@@ -9,12 +9,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
     Friend Partial Class BoundObjectCreationExpression
 
+        Public Sub New(syntax As SyntaxNode, constructorOpt As MethodSymbol, arguments As ImmutableArray(Of BoundExpression), defaultArguments As BitVector, initializerOpt As BoundObjectInitializerExpressionBase, type As TypeSymbol, Optional hasErrors As Boolean = False)
```
>defaultArguments As BitVector [](start = 126, length = 29)

It feels like we should simply add it to the constructor below, as a required parameter

333fred replied:
There are 20 callsites for the constructor below, so I'm going to add it as an optional parameter to avoid bloating this change even more.

---
In reply to: [190019411](https://github.com/dotnet/roslyn/pull/26781#discussion_r190019411) [](ancestors = 190019411)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundPropertyAccess.vb:
```diff
@@ -13,7 +13,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
     Partial Friend Class BoundPropertyAccess
-        Public Sub New(syntax As SyntaxNode, propertySymbol As PropertySymbol, propertyGroupOpt As BoundPropertyGroup, accessKind As PropertyAccessKind, isWriteable As Boolean, receiverOpt As BoundExpression, arguments As ImmutableArray(Of BoundExpression), Optional hasErrors As Boolean = False)
+        Public Sub New(syntax As SyntaxNode, propertySymbol As PropertySymbol, propertyGroupOpt As BoundPropertyGroup, accessKind As PropertyAccessKind, isWriteable As Boolean, isLValue As Boolean, receiverOpt As BoundExpression, arguments As ImmutableArray(Of BoundExpression), type As TypeSymbol, Optional hasErrors As Boolean = False)
```
>Public Sub New(s [](start = 8, length = 16)

Please break this long line

333fred replied:
Going to remove this constructor, no longer relevant.

---
In reply to: [190019739](https://github.com/dotnet/roslyn/pull/26781#discussion_r190019739) [](ancestors = 190019739)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundPropertyAccess.vb:
```diff
@@ -23,10 +27,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
+
+        Public Sub New(syntax As SyntaxNode, propertySymbol As PropertySymbol, propertyGroupOpt As BoundPropertyGroup, accessKind As PropertyAccessKind, isWriteable As Boolean, receiverOpt As BoundExpression, arguments As ImmutableArray(Of BoundExpression), Optional hasErrors As Boolean = False)
+            Me.New(syntax, propertySymbol, propertyGroupOpt, accessKind, isWriteable, receiverOpt, arguments, defaultArguments:=BitVector.Null, hasErrors)
```
>defaultArguments:=BitVector.Null [](start = 110, length = 32)

It is better to not have constructors with assumptions like that, they make it easy to drop important information when not intended.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Attributes.vb:
```diff
@@ -347,7 +347,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                             ' There should not be any used temporaries or copy back expressions because arguments must
                             ' be constants and they cannot be passed byref. 
-                            boundArguments = PassArguments(node.Name, methodResult, boundArguments, diagnostics)
+                            Dim argumentInfo = PassArguments(node.Name, methodResult, boundArguments, diagnostics)
```
>Dim argumentInfo [](start = 28, length = 16)

Please spell out the type

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Invocation.vb:
```diff
@@ -843,7 +844,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End If
 
-            boundArguments = PassArguments(node, bestResult, boundArguments, diagnostics)
+            Dim argumentInfo = PassArguments(node, bestResult, boundArguments, diagnostics)
```
>Dim argumentInfo [](start = 12, length = 16)

Please spell out the type

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Binding/Binder_ObjectInitializer.vb:
```diff
@@ -459,7 +460,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                     Dim methodResult = results.BestResult.Value
 
-                    boundArguments = PassArguments(typeNode, methodResult, boundArguments, diagnostics)
+                    Dim argumentInfo = PassArguments(typeNode, methodResult, boundArguments, diagnostics)
```
>Dim argumentInfo [](start = 20, length = 16)

Please spell out the type

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -1024,14 +1011,7 @@ Namespace Microsoft.CodeAnalysis.Operations
-            ' Some bound nodes used by the boundCaseBlock.CaseStatement.CaseClauses are also going to be used in boundCaseBlock.CaseStatement.ConditionOpt.
-            ' If we simply create another tree based on boundCaseBlock.CaseStatement.ConditionOpt, due to the caching we will end up with the same
-            ' IOperation nodes in two trees, and two parents will compete for assigning itself as the parent - trouble. To avoid that, we simply use
-            ' a new factory to create IOperation tree for the condition. Note that the condition tree is internal and regular consumers cannot get to
```
>BitVector.Null [](start = 87, length = 14)

Why ```BitVector.Null``` is the right value here, aren't all arguments default? Since the entire access is compiler generated and doesn't correspond to anything in the syntax, that is probably fine. Consider adding comment. However, there is a question, why we are not doing the same for GetEnumeratorMethod and MoveNextMethod?

333fred replied:
Note: this comment was for line 1192

---
In reply to: [190031533](https://github.com/dotnet/roslyn/pull/26781#discussion_r190031533) [](ancestors = 190031533)

333fred replied:
We should be passing DefaultArguments through here, will fix.

---
In reply to: [190075484](https://github.com/dotnet/roslyn/pull/26781#discussion_r190075484) [](ancestors = 190075484,190031533)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_Methods.vb:
```diff
@@ -160,6 +160,7 @@ Namespace Microsoft.CodeAnalysis.Operations
             argument As BoundExpression,
             parameters As ImmutableArray(Of VisualBasic.Symbols.ParameterSymbol),
+            ByRef defaultArguments As BitVector,
             invocationWasCompilerGenerated As Boolean
```
>invocationWasCompilerGenerated As Boolean [](start = 12, length = 41)

It looks like this parameter is no longer used.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_Methods.vb:
```diff
@@ -190,7 +190,7 @@ Namespace Microsoft.CodeAnalysis.Operations
                                 kind = ArgumentKind.ParamArray
                             End If
-                        Else
+                        ElseIf Not defaultArguments.IsNull AndAlso defaultArguments(index) Then
```
>Not defaultArguments.IsNull AndAlso defaultArguments(index) [](start = 31, length = 59)

It looks like indexing is safe even for IsNull case. Consider simply passing the boolean instead of passing entire structure by ref.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForEachLoopStatement.vb:
```diff
@@ -2926,7 +2926,7 @@ Block[B1] - Block
                 IParameterReferenceOperation: x (OperationKind.ParameterReference, Type: C) (Syntax: 'x')
               Arguments(1):
-                  IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: member) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: 'x')
+                  IArgumentOperation (ArgumentKind.DefaultValue, Matching Parameter: member) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: 'x')
```
>ArgumentKind.DefaultValue [](start = 38, length = 25)

This change feels unexpected. This entire call is compiler generated, including all the arguments. Otherwise, I would expect the same handling for Current property

333fred replied:
Yeah, I missed properties. Adding.

---
In reply to: [190035385](https://github.com/dotnet/roslyn/pull/26781#discussion_r190035385) [](ancestors = 190035385,190034922)

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.vb:
```diff
@@ -540,6 +578,490 @@ BC30456: 'MissingField' is not a member of 'C'.
+    Class C
+        Implements IEnumerable(Of Integer)
+
+        Sub M(a, b)
```
>a, b [](start = 14, length = 4)

Please spell out the types

AlekseyTs commented on src/Tools/Source/CompilerGeneratorTools/Source/BoundTreeGenerator/BoundNodeClassWriter.cs:
```diff
@@ -86,6 +86,7 @@ private void InitializeValueTypes()
             _valueTypes.Add("ImmutableArray", true);
             _valueTypes.Add("PropertyAccessKind", true);
+            _valueTypes.Add("BitVector", true);
         }
```
We probably shouldn't hard code it here, see my comments in BoundNodes.xml

333fred commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundCall.vb:
```diff
@@ -19,13 +39,31 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+            Me.New(syntax, method, methodGroupOpt, receiverOpt, arguments, defaultArguments:=BitVector.Null, constantValueOpt, isLValue, suppressObjectClone, type, hasErrors)
+        End Sub
+
+        Public Function Update(method As MethodSymbol, methodGroupOpt As BoundMethodGroup, receiverOpt As BoundExpression, arguments As ImmutableArray(Of BoundExpression), constantValueOpt As ConstantValue, isLValue As Boolean, suppressObjectClone As Boolean, type As TypeSymbol) As BoundCall
```
12 currently.

---
In reply to: [190018611](https://github.com/dotnet/roslyn/pull/26781#discussion_r190018611) [](ancestors = 190018611)

AlekseyTs replied:
> 12 currently.

What worries me about a helper like that is that it allows to switch method, arguments. etc. Making it very easy to get DefaultArguments out of sync. I would recommend to not introduce a helper like that. It is better to add non-optional parameter to existing APIs

---
In reply to: [190037090](https://github.com/dotnet/roslyn/pull/26781#discussion_r190037090) [](ancestors = 190037090,190018611)

AlekseyTs commented:
Done with review pass (iteration 6)

333fred commented:
This was an error with the merge, which is interesting because it didn't conflict.

---
In reply to: [391088751](https://github.com/dotnet/roslyn/pull/26781#issuecomment-391088751) [](ancestors = 391088751)

333fred commented:
@AlekseyTs addressed feedback

---
In reply to: [391126791](https://github.com/dotnet/roslyn/pull/26781#issuecomment-391126791) [](ancestors = 391126791)

AlekseyTs commented:
@333fred Build is failing:
error CS0612: 'OperationKind.CollectionElementInitializer' is obsolete [D:\j\workspace\windows_debug---73b70dbc\src\Test\Utilities\Portable\TestUtilities.csproj]
Indication 1

AlekseyTs commented on src/Compilers/VisualBasic/Portable/BoundTree/BoundObjectCreationExpression.vb:
```diff
@@ -9,12 +9,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
         Public Function Update(constructorOpt As MethodSymbol, arguments As ImmutableArray(Of BoundExpression), initializerOpt As BoundObjectInitializerExpressionBase, type As TypeSymbol) As BoundObjectCreationExpression
-            Return Update(constructorOpt, Nothing, arguments, initializerOpt, type)
+            Return Update(constructorOpt, Nothing, arguments, DefaultArguments, initializerOpt, type)
```
>DefaultArguments [](start = 62, length = 16)

Can we pass it explicitly?

333fred replied:
I missed removing this function, will remove altogether.

---
In reply to: [190109952](https://github.com/dotnet/roslyn/pull/26781#discussion_r190109952) [](ancestors = 190109952)

333fred replied:
Scratch that, quick glance missed that this was preexisting. I'll fully-qualify this and the `Nothing` arg.

---
In reply to: [190380797](https://github.com/dotnet/roslyn/pull/26781#discussion_r190380797) [](ancestors = 190380797,190109952)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_Methods.vb:
```diff
@@ -180,22 +180,11 @@ Namespace Microsoft.CodeAnalysis.Operations
 
-                    If argument.WasCompilerGenerated AndAlso Not invocationWasCompilerGenerated Then
-
-                        If index = lastParameterIndex AndAlso ParameterIsParamArray(parameters(lastParameterIndex)) Then
```
>index = lastParameterIndex [](start = 27, length = 26)

It looks like this check is dropped, this doesn't look safe.

333fred replied:
The new check is precisely the check used by other sections of the compiler to determine if the parameter is a paramarray. I've checked all instances of creating and updating BoundArrayCreation, and the only time that flag is set is here: [https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Binding/Binder_Invocation.vb#L2705](https://github.com/dotnet/roslyn/blob/master/src/Compilers/VisualBasic/Portable/Binding/Binder_Invocation.vb#L2705).

---
In reply to: [190111038](https://github.com/dotnet/roslyn/pull/26781#discussion_r190111038) [](ancestors = 190111038)

333fred replied:
I can reintroduce the index check if you want, but this is what `BoudnArrayCreation.IsParamArrayArgument` is intended to convey.

---
In reply to: [190316564](https://github.com/dotnet/roslyn/pull/26781#discussion_r190316564) [](ancestors = 190316564,190111038)

333fred replied:
Talked with Aleksey offline, this check is sufficient.

---
In reply to: [190316977](https://github.com/dotnet/roslyn/pull/26781#discussion_r190316977) [](ancestors = 190316977,190316564,190111038)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory_Methods.vb:
```diff
@@ -180,22 +180,11 @@ Namespace Microsoft.CodeAnalysis.Operations
-                            '       a compiler generated argument for a ParamArray parameter is created iff
-                            '       a list of arguments (including 0 argument) is provided for ParamArray parameter in source
-                            '       https://github.com/dotnet/roslyn/issues/18550
-                            If argument.Kind = BoundKind.ArrayCreation Then
```
>If argument.Kind = BoundKind.ArrayCreation Then [](start = 28, length = 47)

It looks like this check is dropped, this doesn't look safe.

AlekseyTs commented:
Done with review pass (iteration 8)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [391206536](https://github.com/dotnet/roslyn/pull/26781#issuecomment-391206536) [](ancestors = 391206536)

AlekseyTs commented:
Compile Error

error BC30516: Overload resolution failed because no accessible 'Update' accepts this number of arguments. [/mnt/j/workspace/dotnet_roslyn/features_dataflow/ubuntu_16_debug_prtest/src/Compilers/VisualBasic/Portable/BasicCodeAnalysis.vbproj]
Indication 1

jaredpar commented on src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:
```diff
@@ -1,23 +1,23 @@
     /// This interface is reserved for implementation by its associated APIs. We reserve the right to
     /// change it in the future.
     /// </remarks>
+    [Obsolete("ICollectionElementInitializerOperation has been replaced with IInvocationOperation and IDynamicInvocationOperation", error: true)]
```
>has [](start = 54, length = 3)

Use nameof here?

333fred replied:
Used nameof everywhere I could. I couldn't use nameof(ICollectionElementInitializer) in the attributes on the visitors/OperationKind, so I kept the plain text there but still swapped IInvocationOperation and IDynamicInvocationOperation for nameof.

---
In reply to: [190395155](https://github.com/dotnet/roslyn/pull/26781#discussion_r190395155) [](ancestors = 190395155)

jaredpar commented:
>         /// Add method invoked on collection. Null for dynamic invocation and error cases.

This still mentions dynamic. 

---
Refers to: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:24 in 27dbf2e. [](commit_id = 27dbf2ebb5e0534a7df78a7f4f6ae10dc56905e7, deletion_comment = False)

jaredpar commented:
>         /// Add method invoked on collection. Null for dynamic invocation and error cases.

Given the type is obsolete and no longer used anywhere should we even keep the member comments? Seems like it would just cause confusion at this point. But I'm not sure what our policy is here. 

---
Refers to: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:24 in 27dbf2e. [](commit_id = 27dbf2ebb5e0534a7df78a7f4f6ae10dc56905e7, deletion_comment = False)

333fred commented:
>         /// Add method invoked on collection. Null for dynamic invocation and error cases.

Based on feedback from the email chain, I'll remove these comments.

---
In reply to: [391498240](https://github.com/dotnet/roslyn/pull/26781#issuecomment-391498240) [](ancestors = 391498240)

---
Refers to: src/Compilers/Core/Portable/Operations/ICollectionElementInitializerOperation.cs:24 in 27dbf2e. [](commit_id = 27dbf2ebb5e0534a7df78a7f4f6ae10dc56905e7, deletion_comment = False)

jaredpar commented on src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs:
```diff
@@ -1313,17 +1313,12 @@ public override void VisitMemberInitializer(IMemberInitializerOperation operatio
             Visit(operation.Initializer, "Initializer");
         }
 
+        [Obsolete("ICollectionElementInitializerOperation has been replaced with IInvocationOperation and IDynamicInvocationOperation", error: true)]
```
Use nameof

jaredpar replied:
NM. Saw your other comment. 

---
In reply to: [190641336](https://github.com/dotnet/roslyn/pull/26781#discussion_r190641336) [](ancestors = 190641336)

333fred commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs:
```diff
@@ -113,6 +113,16 @@ private Lazy<IOperation> CreateReceiverOperation(BoundNode instance, ISymbol sym
+        }
+
+        private bool IsMethodInvalid(LookupResultKind resultKind, MethodSymbol targetMethod) =>
+            resultKind == LookupResultKind.OverloadResolutionFailure || targetMethod == null || targetMethod.OriginalDefinition is ErrorMethodSymbol;
```
Will do in a followup PR.

---
In reply to: [190639853](https://github.com/dotnet/roslyn/pull/26781#discussion_r190639853) [](ancestors = 190639853)

333fred commented:
@mavasani for review please.
